# Montu Python
## Astronomical ephemerides for the ancient world

<p align="center"><img src="https://github.com/seap-udea/MontuPython/raw/main/montu/data/montu-python-logo-complete.webp" width="300" alt="MontuPython logo"/></p>

[![PyPI](https://img.shields.io/pypi/v/montu?color=blue)](https://pypi.org/project/montu/) [![Downloads](https://img.shields.io/pypi/dw/montu)](https://pypi.org/project/montu/) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) [![Python](https://img.shields.io/badge/python-3.8%20|%203.9%20|%203.10%20|%203.11-blue)](https://www.python.org/) [![GitHub](https://img.shields.io/badge/GitHub-seap--udea%2FMontuPython-blue?logo=github)](https://github.com/seap-udea/MontuPython) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.21753938.svg)](https://doi.org/10.5281/zenodo.21753938) [![arXiv](https://img.shields.io/badge/arXiv-preprint%20forthcoming-b31b1b)](#) [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/README.ipynb)
[![Powered by PyEphem](https://img.shields.io/badge/powered%20by-PyEphem-blue)](https://pypi.org/project/ephem/) [![Powered by PyMeeus](https://img.shields.io/badge/powered%20by-PyMeeus-blue)](https://pypi.org/project/pymeeus/) [![Powered by spiceypy](https://img.shields.io/badge/powered%20by-spiceypy-blue)](https://pypi.org/project/spiceypy/) [![Powered by pyplanets](https://img.shields.io/badge/powered%20by-pyplanets-blue)](https://pypi.org/project/pyplanets/)

`MontuPython` is a Python package intended to compute astronomical ephemerides in the ancient world, thousands of years before present. It was initially designed to compute ephemerides for the ancient Egypt, but it can also be used to study astronomical phenomena in other sites of interest for cultural astronomy (archeoastronomy).

The package was originally developed to streamline calculations on archeoastronomy that were previously performed manually using astronomical software such as Stellarium. While there are already outstanding libraries for positional astronomy—developed by a vibrant community of professionals and enthusiasts—such as [PyEphem](https://pypi.org/project/ephem/), [pyplanets](https://pypi.org/project/pyplanets/), [PyMeeus](https://pypi.org/project/pymeeus/), [spiceypy](https://pypi.org/project/spiceypy/), and others (many of which make MontuPython possible)—our goal here is different. We aim to provide a tool that not only simplifies interaction with these libraries, but, more importantly, is specialized for the study of astronomical phenomena that occurred thousands of years ago.

While MontuPython places a special emphasis on ancient Egypt—understandably, since Egyptian religion, culture, and statehood were profoundly intertwined with astronomy—the tool is broadly applicable to many other contexts. In fact, MontuPython can be used to study archaeoastronomical phenomena in any culture around the world. Its features are designed to support research across a diverse range of civilizations, making it a versatile resource for all archaeologists, historians, and cultural astronomers interested in humanity's relationship with the sky.

MontuPython has benefited from the contributions and advice of several collaborators:

- Francisco "Tito" Vivas provided an actual archaeoastronomical fase of study who originally motivated the development of this software.
- Prof. José Lull of the Universitá Autónoma de Barcelona (UAB), provided scientific, archaeoastronomical, and Egyptological advice.
- Juanita Agudelo developed the first versions of the MontuPython web app.
- Luis Arroyo developed the version of the web app which precedes the desktop app.

## Useful Resources 

For more information before you begin, please refer to:

- **MontuPython documentation**: The full documentation of the package is available at [https://montupython.readthedocs.io](https://montupython.readthedocs.io/en/latest/).

- **MontuPython Desktop**: This is the graphical front-end to the library, built with PySide6. It offers pre-built tools for users who prefer not to program at the Python level, or who want quick access to core functionality — including a **Calendar Calculator** and **Seasons & Lunar Phases** module.

  Pre-built installers for **macOS** and **Windows** are published on **[GitHub Releases](https://github.com/seap-udea/MontuPython/releases)**. Open that page and look for a release whose tag starts with `desktop-v` (for example, `desktop-v0.1.1`). Download the file that matches your operating system:

    | Platform | File to download | How to run |
    |----------|------------------|------------|
    | **[Descarga versión de macOS](https://jorgezuluaga.github.io/sh/montupython-desktop-macos)** | `*-macos.dmg` or `*-macos.zip` | Open the `.dmg` and drag **MontuPython Desktop** to Applications, or unzip the `.zip` and open `MontuPython Desktop.app`. If macOS shows a security warning the first time, right-click the app and choose **Open**. |
    | **[Descarga versión de Windows](https://jorgezuluaga.github.io/sh/montupython-desktop-windows)** | `*-windows.zip` | Unzip the archive, open the `MontuPython-Desktop` folder, and double-click `MontuPython-Desktop.exe`. If SmartScreen warns you, choose **More info** → **Run anyway**. |

    No Python installation is required. Most features work offline; the observer map (OpenStreetMap) needs an internet connection.
    
    If you already use the `montu` library, install the optional **Desktop** extra and launch the GUI from a terminal:

    ```bash
    pip install montu[desktop]
    montu-gui
    # or
    imontu --gui
    ```

    `montu[desktop]` adds **PySide6**, **Pygments**, and **plotly**. On first launch, `montu-gui` / `imontu --gui` also fetches the `montu_gui` sources if they are not already present in your checkout. The same extra is required for `imontu --sothic`.

## Download and install

### Library only

Install the core package from PyPI:

```bash
pip install montu
```

This pulls in the pinned astronomy stack (`ephem`, `pymeeus`, `pyplanets`), Dash (MontuApp), and the usual scientific Python dependencies. It does **not** install PySide6 or other MontuPython Desktop GUI packages.

### Library + Desktop GUI

Add the optional **Desktop** extra when you want the PySide6 GUI or `imontu --sothic`:

```bash
pip install montu[desktop]
```

From a development checkout:

```bash
pip install -e ".[desktop]"
```

## Usage in Google Colab

Uncomment and run in the next cell (replace the branch name if needed), then **restart the runtime** before importing `montu`:

In [1]:
# Uncomment if you are in Google Colab
# !pip install -q "git+https://github.com/seap-udea/MontuPython@maat"
# For Desktop GUI in a local Python environment (not typical on Colab):
# !pip install -q "montu[desktop] @ git+https://github.com/seap-udea/MontuPython@maat"

## Tutorial by example notebooks

The following notebooks introduce MontuPython step by step. Open any of them in Google Colab and start working immediately — they are ideal for training new users or for testing code generated by MontuPython Desktop.


- **Code snippets** Runnable examples — one MontuPython feature per snippet, each with a short explanation. Covers utilities, time and calendars, observers, stars, Sun/Moon/planets, seasons, sky maps, heliacal rises, solar eclipses, conjunctions, and planetary stations. A good first overview before the topic-specific notebooks below.

   **File**: `MontuPython-CodeSnippets.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-CodeSnippets.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-CodeSnippets.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-CodeSnippets.ipynb)

- **Time: Basic time functionalities** Illustrates the `montu.Time` class across its full range of calendars and time scales. Parse dates in ISO, SPICE, Julian day, and terrestrial time; convert between proleptic Gregorian, mixed Julian/Gregorian, ancient Egyptian civil, and sothic (Horus year) calendars; and add or subtract calendar periods.

   **File**: `MontuPython-MonTime.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-MonTime.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-MonTime.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-MonTime.ipynb)

- **Egyptian civil calendar: historical cross-checks** Compares MontuPython's computed Egyptian civil dates against the civil dates recorded in Egyptology sources for key historical events (apokatastasis anchors, Ptolemaic decrees, dated papyri). Uses the same event catalogue as MontuPython Desktop's Calendar Calculator.

   **File**: `MontuPython-EgyptianCalendar.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-EgyptianCalendar.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-EgyptianCalendar.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-EgyptianCalendar.ipynb)

## Advanced examples


The following notebooks present archaeoastronomical computations of broader historical and interpretive interest.


- **Heliacal rises: Sirius step by step** Reconstructs the first morning visibility of Sirius from Thebes around 2782 BCE — the era of the first *apokatastasis* of the Sothic cycle. Works through all four heliacal-rise models in MontuPython (Ptolemy, Schaefer 1985/1987, Belokrylov 2011) and explains each visibility criterion rather than treating `HeliacalRise` as a black box.

   **File**: `MontuPython-HeliacalRises.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-HeliacalRises.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-HeliacalRises.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-HeliacalRises.ipynb)

- **Solar eclipses** Predicts **local circumstances** of solar eclipses from the bundled NASA Five Millennium catalogue (Espenak & Meeus): search the catalogue, validate the 2024 Dallas totality, and survey historical eclipses visible from Thebes.

   **File**: `MontuPython-SolarEclipses.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-SolarEclipses.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-SolarEclipses.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-SolarEclipses.ipynb)

- **Conjunctions** Finds angular conjunctions of planets and stars with `Conjunction` and `ConjunctionExplorer`: separation and visibility at one epoch, lapse intervals, Plotly sky maps, and reference cases from simple pairs to planetary trios and retrograde triple crossings.

   **File**: `MontuPython-Conjunctions.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-Conjunctions.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-Conjunctions.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-Conjunctions.ipynb)

- **Montunctions** Computes *marstices* — the stationary points of Mars — and their conjunctions with Aldebaran near the winter solstice throughout ancient Egyptian history. Includes periodicity analysis and publication-ready plots. The methodology and figures supported Francisco Vivas Fernández's Ph.D. thesis on the astronomical and landscape orientations of Senenmut's monuments in the Theban necropolis (2023).

   **File**: `MontuPython-Montunctions.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-Montunctions.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-Montunctions.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-Montunctions.ipynb)

- **Example: Venus Azimutal Distributions** Computes the distribution of Venus rising and setting azimuths at dawn and dusk over millennia, as observed from Thebes. Distinguishes morning-star and evening-star apparitions and summarizes their preferred horizon directions — a classic archaeoastronomical exercise for Venus cult and calendar studies.

   **File**: `MontuPython-VenusAzimuths.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-VenusAzimuths.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-VenusAzimuths.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-VenusAzimuths.ipynb)

- **Observer Horizons** Computes and plots the real topographical horizon profile for an observer by downloading and processing Copernicus GLO-30 DEM tiles. Demonstrates interactive Plotly and Leaflet maps, celestial object occultation, and sunrise azimuth calculations.

   **File**: `MontuPython-ObserverHorizon.ipynb`. **Links**: [GitHub](https://github.com/seap-udea/MontuPython/blob/main/examples/MontuPython-ObserverHorizon.ipynb) | [ReadTheDocs](https://montupython.readthedocs.io/en/latest/examples/MontuPython-ObserverHorizon.html) | [Colab](https://colab.research.google.com/github/seap-udea/MontuPython/blob/main/examples/MontuPython-ObserverHorizon.ipynb)


## What's new


Release notes are now maintained in [WHATSNEW.md](WHATSNEW.md).

<p align="center"><img src="https://github.com/seap-udea/MontuPython/blob/main/gallery/peret-sopedet-illustration.webp?raw=true" alt="Illustration of the heliacal rise of Sirius (Peret Sopedet) on the Giza plateau" width="800"/></p>

<p align="center"><em>Illustration of Egyptian Peret Sopedet on the Giza plateau</em> — the heliacal rise of Sirius (<i>Sopedet</i>) that marks the civil New Year (<code>[hrw 0] I <i>akhet</i> 1</code>). Image generated with Nano Banana, edited in Preview, based on Bob Moler’s Stellarium–GIMP simulation (<a href="https://bobmoler.blog/2017/02/07/02072017-ephemeris-sirius-an-important-star-in-history/">Bob Moler's Ephemeris Blog</a>, 7 February 2017).</p>


## AI assistance disclosure

Starting with version *0.20.0*, portions of the code, inline documentation, code review, and debugging in this repository were assisted by AI language models, including **Cursor Composer 2.5**, **Anthropic Claude Sonnet 4.6**.

The human authors maintain that all scientific ideas, the overall project conception, the package and notebook architecture, the design of the numerical and scientific workflows, and their interpretation remain original contributions of the human authors. AI tools were used exclusively as coding and writing assistants — analogous to a spell-checker or a compiler — and bear no authorship over the scientific content of this project. AI models also assisted with translating text from Spanish (the native language of the human authors) into English, and with English spelling and grammar review.

## Quickstart

In the rest of this document we show you basic applications of the package. Since how it is imported until how it can be used for realistic calculations.

This README is at the same time an executable notebook. A runable version of the README can be find in the [Google Drive Repo of the package](https://drive.google.com/drive/folders/11L59yZ3A1g1ZT7v_dLDPwLMnRMR-tFgE?usp=sharing) as the file [`README.ipynb`]() 

### Preparation

> **NOTE:** This notebook is autogenerated with `Jupyter`. Some of the commands you will find here are only intended to make the README executable. In no sense are mandatory when using `MontuPython`.

You may import the package using:

In [2]:
%matplotlib inline
import matplotlib.pyplot as plt
!mkdir -p gallery

import montu

MontuPython version 0.50.1. 𓇍𓇋𓇋𓏏𓅓𓊵 𓎛𓎡𓄿𓀭𓎛𓈖𓂝𓎡 (ii-ti m Htp, HkAx Hn'-k)


It is important that before using the most interesting commands of the package, load relevant data:

In [3]:
# Useful aliases
from montu import D2S, PRINTDF, TABLEDF
# Load stars
allstars = montu.Stars()

Loading stellar catalogue montu_stellar_catalogue_v38.csv


Two very basic codes that do something non-trivial albeit simple calculations in `MontuPython are:`

1. Compute the position of mars at a given time and while observing from a given site on Earth:

In [4]:
mtime = montu.Time('-2500-01-01 12:00:00.00')
tebas = montu.Observer(lon=33,lat=24)
mars = montu.Planet('Mars')
mars.conditions_in_sky(at=mtime,observer=tebas)
mars.show_conditions()

Mars — sky conditions
  Epoch: -2500-01-22 12:00:00 / -2500-01-01 12:00:00.000000  (JED 807954.000000)
  Site: Custom site — lat 24.000000°, lon 33.000000°, 0 m  (P=1013.25 mbar, T=15.0 °C)
  Name: Mars
  Hour angle: 12:20:15.604 h
  Visual magnitude: -1.13 mag
  Rise time (UTC): -2500-01-22 16:49:49
  Rise azimuth: 63.133665°
  Set time (UTC): -2500-01-23 06:24:24
  Set azimuth: 296.924743°
  Transit time (UTC): -2500-01-22 23:37:37
  Transit elevation: 89.841412°
  Elongation from Sun: -157.818359°
  Distance from Earth: 0.660488 AU
  Distance from Sun: 1.626107 AU
  Angular diameter: 14.171"
  Illuminated fraction: 98.65 %
  Heliocentric latitude: 1.888422°
  Heliocentric longitude: 111.258645°
  Heliocentric longitude (alt.): 111.258645°
  Circumpolar: no
  Never rises: no


2. Obtain the information about a star from the stellar catalogue and, as in the case of the planet, obtain the position of the star in the sky.

In [5]:
mtime = montu.Time('-2500-01-01 12:00:00.00')
tebas = montu.Observer(lon=33,lat=24,height=0)
aldebaran = allstars.get_stars(ProperName='Aldebaran', return_as='Star')
aldebaran.where_in_sky(at=mtime,observer=tebas)
aldebaran.show_position()

Aldebaran — sky position
  Epoch: -2500-01-22 12:00:00 / -2500-01-01 12:00:00.000000  (JED 807954.000000)
  Site: Custom site — lat 24.000000°, lon 33.000000°, 0 m  (P=1013.25 mbar, T=15.0 °C)
  Name: Aldebaran
  RA (J2000): 04:35:35.599 h
  Dec (J2000): 16.745934°
  RA (J2000, proper motion): 04:35:36.404 h
  Dec (J2000, proper motion): 16.745996°
  RA (epoch): 00:36:32.065 h
  Dec (epoch): -2.261722°
  RA (geocentric): 00:36:32.065 h
  Dec (geocentric): -2.261722°
  Azimuth: 107.617963°
  Elevation: 29.633599°


### Working with time

One of the most interesting and basic functionalities of MontuPython is to convert date among 
different type of calendars and astronomical scales.  You may taste these functionalities using:

In [6]:
mtime = montu.Time('bce2501-01-01 12:00:00')
mtime

Time('-2500-01-01 12:00:00.000000'/'-2500-01-22 12:00:00'/'[hrw 280] I shemu 17'/JED 807954.0/JTD 807954.6909688)

other alternative formats for the same date are:

In [7]:
mtime = montu.Time('2501 b.c.e. 01-01 12:00:00')
mtime = montu.Time('-2500-01-01 12:00:00')

If you print this time object you will get:

In [8]:
print(mtime)

2501 B.C. 01-01 12:00:00.00 / [hrw 280] I shemu 17:
    Date in ISO format: 2501 B.C. 01-01 12:00:00.00
    Date in proleptic UTC: -2500-01-01 12:00:00.000000
    Date in mixed UTC: -2500-01-22 12:00:00
    Weekday: 2 (monday)
    Date in sothic format: [hrw 280] I shemu 17
    Terrestrial time: tt [seconds]: -142006202700.3
    UTC time: jed [days]: 807954.0
    Delta-t = TT - UTC [seconds]: 59699.7



For more details:

In [9]:
mtime.details()

Montu Time Object:
-------------------------- 
Readable:
    Date in proleptic UTC (.readable.datepro): -2500-01-01 12:00:00.000000
    Date in mixed UTC (.readable.datemix): -2500-01-22 12:00:00
    Date in SPICE format (.readable.datespice): 2501 B.C. 01-01 12:00:00.00
    Date in sothic format (.readable.datesot): [hrw 280] I shemu 17
    Weekday (.readable.weekday): 2 (monday)
    Components (.readable.comps): [-1, 2500, 1, 1, 12, 0, 0, 0]
Objects:
    Date in datetime64 format (.readable.obj_datetime64): -2500-01-01T12:00:00.000000
    Date in PyPlanet Epoch (.obj_pyplanet): 807954.0
    Date in PyEphem Epoch (.obj_pyephem): -2501/1/22 12:00:00
General:
    Is bce (.bce): True
    Is Julian (.isjulian): True
Uniform scales:
    Terrestrial time:
        tt (.tt): -142006202700.3
        jtd (.jtd): 807954.6909688
        htd (.htd): 102457.19096879999
    UTC time:
        et (.et): -142006262400.0
        jed (.jed): 807954.0
        hed (.hed): 102456.5
    Delta-t = TT - UTC (.

Notice that the date in Gregorian proleptic will be **bce 2501-01-01** but in the mixed calendar that uses Julian calendar before its adoption at 1582-10-04, will be **bce 2501-01-22**.

#### Egyptian civil calendar

In the previous output you may also notice that the class `montu.Time` automatically convert the gregorian/julian date into the *civil egyptian calendar* or the *Sothic* calendar. In the previous example, the date bce 2501-01-22 (julian) correspond to the civil egyptian date of **I *shemu* 17** (MontuPython: `[hrw 280] I shemu 17`). Since the years in the *Sothic calendar* were regularly referred to specific king's reigns, we have introduced in `MontuPython` the so-called *Horus years* (abreviated *hrw*). The zero *Horus year* is bce 2782 which corresponds to the first *apokatastasis*, namely the year when the day **I *akhet* 1** coincide with the heliacal rise of Sirius. 

You may also provide a date in the *Sothic calendar* and obtain the corresponding julian date:

In [10]:
mtime = montu.Time('[hrw 1461] I akhet 1',calendar='sothic')
print(mtime)

1322 B.C. 07-20 08:45:50.800000 / [hrw 1461] I akhet 1:
    Date in ISO format: 1322 B.C. 07-20 08:45:50.800000
    Date in proleptic UTC: -1321-07-20 08:45:50.8
    Date in mixed UTC: -1321-07-20 00:00:00
    Weekday: 3 (tuesday)
    Date in sothic format: [hrw 1461] I akhet 1
    Terrestrial time: tt [seconds]: -104784376449.2
    UTC time: jed [days]: 1238762.5
    Delta-t = TT - UTC [seconds]: 31550.8



As you may see, this is the second *apokatastasis*, ie. in bce 1322-07-20, the heliacal rise of Sirius happened again at I *akhet* 1.

#### Operations with dates

You may add or substract time to a given date. This is done by adding or substracting seconds to the reference time:

In [11]:
mtime = montu.Time('2001-01-01 12:00:00',format='iso')
(mtime,
 mtime - 12*montu.HOUR, 
 mtime + 1*montu.DAY, 
 mtime - 3*montu.CALYEAR, 
 mtime + 20*montu.JULYEAR)

(Time('2001-01-01 12:00:00.000000'/'2001-01-01 12:00:00'/'[hrw 4784] I shemu 14'/JED 2451911.0/JTD 2451911.0007419),
 Time('2001-01-01 00:00:00.000000'/'2001-01-01 00:00:00'/'[hrw 4784] I shemu 14'/JED 2451910.5/JTD 2451910.5007419),
 Time('2001-01-02 12:00:00.000000'/'2001-01-02 12:00:00'/'[hrw 4784] I shemu 15'/JED 2451912.0/JTD 2451912.0007419),
 Time('1998-01-02 12:00:01.097278'/'1998-01-02 12:00:00'/'[hrw 4781] I shemu 14'/JED 2450816.0000127/JTD 2450816.0007419),
 Time('2021-01-01 11:59:51.904328'/'2021-01-01 11:59:59'/'[hrw 4804] I shemu 19'/JED 2459215.9999063/JTD 2459216.0007419))

As you may notice, adding or substracting and integer number of seconds not necesarily correspond to adding or sutracting days or years to the calendar. This is because of the difference in UT and TT: UT is always behind TT by a certain amount of seconds.  Normally leapseconds are included every once in a while. However to calculate ephemerides in the ancient world, `MontuPython` uses a continuous model of deltat that small discrepancies from year to year.

To add calendar periods of time you may use the `add` method of `Montu.Time`:

In [12]:
mtime = montu.Time('2001-01-01 12:00:00')
mtime2 = mtime.add(years=1)
print(mtime2)

2002-01-01 12:00:00.000000 / [hrw 4785] I shemu 14:
    Date in ISO format: 2002-01-01 12:00:00.000000
    Date in proleptic UTC: 2002-01-01 12:00:00.000000
    Date in mixed UTC: 2002-01-01 12:00:00
    Weekday: 3 (tuesday)
    Date in sothic format: [hrw 4785] I shemu 14
    Terrestrial time: tt [seconds]: 63158464.3
    UTC time: jed [days]: 2452276.0
    Delta-t = TT - UTC [seconds]: 64.3



### Working with stars

The stellar catalogue included with `MontuPython` contains more than 119 000 stars, including almost 9 000 visible to the naked eye.  Information about stars is stored in a `pandas Data Frame` whose columns are:

In [13]:
allstars.data.columns

Index(['MN', 'HD', 'HR', 'HIP', 'Gl', 'Name', 'OtherDesignations',
       'ProperName', 'Bayer', 'Flamsteed', 'Constellation', 'RAJ2000',
       'DecJ2000', 'GalLonJ2000', 'GalLatJ2000', 'pmRA', 'pmDec', 'RadVel',
       'Distance', 'Vmag', 'Vmag_min', 'Vmag_max', 'B-V', 'SpType',
       'Luminosity', 'XJ2000', 'YJ2000', 'ZJ2000', 'VXJ2000', 'VYJ2000',
       'VZJ2000', 'Primary', 'MultipleID', 'IsMultiple', 'IsVariable'],
      dtype='str')

Altough you may manipulate this DataFrame using the conventional commands in pandas, we have designed several useful methods to obtain subsets of the catalogue. For instance if you want to extract the stars visible to naked eye, the command would be:

In [14]:
stars = allstars.get_stars(Vmag=[-2,6.5])
print(f"There is {stars.number} visible to the naked eye in the catalogue")

There is 8920 visible to the naked eye in the catalogue


You can use any of the properties of the stars to filter them. A common filter is to look for single stars:

In [15]:
aldebaran = stars.get_stars(ProperName='Aldebaran', return_as='Star')
aldebaran

Star('Aldebaran'/'Tau'/04:35:55.237/16:30:33.484/V=0.87 mag/d=20.43 pc/Sp=K5III)

To get the whole properties of a Star:

In [16]:
aldebaran.show_properties()

Aldebaran — catalogue properties
  Proper name: Aldebaran
  Name: Aldebaran
  Bayer designation: α Tau
  Flamsteed number: 87 Tau
  Constellation: Tau
  HIP: 21421
  HD: 29139
  HR: 1457
  Gliese: Gl 171.1A
  Other designations: 87 Tau/87Alp Tau/Aldebaran/Gl 171.1A/HD 29139/HIP 21421/HR 1457/HYG 21368/MN 15/α Tau
  RA (J2000): 04:35:55.237 h
  Dec (J2000): 16:30:33.484°
  Galactic longitude: 180:58:18.868°
  Galactic latitude: -20:14:53.892°
  Proper motion RA: 62.78 mas/yr
  Proper motion Dec: -189.36 mas/yr
  Radial velocity: 54.5 km/s
  Distance: 20.4332 pc
  Visual magnitude: 0.87 mag
  B−V colour index: 1.538
  Spectral type: K5III
  Luminosity: 163.23 L☉
  Multiple system: yes
  Variable star: yes


All information about a star is stored in the `data` attribute:

In [17]:
aldebaran = stars.get_stars(ProperName='Aldebaran')
TABLEDF(aldebaran.data)

|    |   MN |    HD |   HR |   HIP | Gl        | Name      | OtherDesignations                                                                     | ProperName   | Bayer   | Flamsteed   | Constellation   |   RAJ2000 |   DecJ2000 |   GalLonJ2000 |   GalLatJ2000 |   pmRA |   pmDec |   RadVel |   Distance |   Vmag |   Vmag_min |   Vmag_max |   B-V | SpType   |   Luminosity |   XJ2000 |   YJ2000 |   ZJ2000 |   VXJ2000 |   VYJ2000 |   VZJ2000 |   Primary | MultipleID   |   IsMultiple |   IsVariable |
|----|------|-------|------|-------|-----------|-----------|---------------------------------------------------------------------------------------|--------------|---------|-------------|-----------------|-----------|------------|---------------|---------------|--------|---------|----------|------------|--------|------------|------------|-------|----------|--------------|----------|----------|----------|-----------|-----------|-----------|-----------|--------------|--------------|--------------

> **NOTE:** The `TABLEDF` is a special function of `MontuPython` that allows you to produce a readable table out of a pandas DataFrame, a list of dictionaries and other iterable python objects. See `tabulate` package for details.

Another useful method included with the class `Stars` is that of filtering the getting the stars close to a given point in the sky. For illustrare, below is the command to obtain all stars in the sky with magnitudes less than 5 and that are at 5.5 degrees or less than Aldebaran:

In [18]:
hyades = stars.get_stars_around(center=[aldebaran.data.RAJ2000,aldebaran.data.DecJ2000],radius=5.5,Vmag=[-1,5])
hyades

18 star(s):
|      |   MN |    HD |   HR |   HIP | Gl        | Name           | OtherDesignations                                                                     | ProperName     | Bayer   | Flamsteed   | Constellation   |   RAJ2000 |   DecJ2000 |   GalLonJ2000 |   GalLatJ2000 |   pmRA |   pmDec |   RadVel |   Distance |   Vmag |   Vmag_min |   Vmag_max |   B-V | SpType   |   Luminosity |   XJ2000 |   YJ2000 |   ZJ2000 |    VXJ2000 |    VYJ2000 |    VZJ2000 |   Primary | MultipleID   |   IsMultiple |   IsVariable |
|------|------|-------|------|-------|-----------|----------------|---------------------------------------------------------------------------------------|----------------|---------|-------------|-----------------|-----------|------------|---------------|---------------|--------|---------|----------|------------|--------|------------|------------|-------|----------|--------------|----------|----------|----------|------------|------------|------------|-----------|--------

We can map the stars:

In [19]:
fig,axs = hyades.plot_stars()
fig.savefig('gallery/hyades.png')
plt.close(fig) # Used only for README generation


<p align="center"><img src="https://github.com/seap-udea/MontuPython/blob/main/gallery/hyades.png?raw=true" alt="Logo""/></p>

Now you can precess the stars:

In [20]:
mtime = montu.Time('-2500-01-01 12:00:00.00')
hyades.where_in_space(at=mtime,inplace=True)
hyades

18 star(s):
|      |   MN |    HD |   HR |   HIP | Gl        | Name           | OtherDesignations                                                                     | ProperName     | Bayer   | Flamsteed   | Constellation   |   RAJ2000 |   DecJ2000 |   GalLonJ2000 |   GalLatJ2000 |   pmRA |   pmDec |   RadVel |   Distance |   Vmag |   Vmag_min |   Vmag_max |   B-V | SpType   |   Luminosity |   XJ2000 |   YJ2000 |   ZJ2000 |    VXJ2000 |    VYJ2000 |    VZJ2000 |   Primary | MultipleID   |   IsMultiple |   IsVariable |           tt |    jed |   RAJ2000t |   DecJ2000t |   RAEpoch |   DecEpoch |
|------|------|-------|------|-------|-----------|----------------|---------------------------------------------------------------------------------------|----------------|---------|-------------|-----------------|-----------|------------|---------------|---------------|--------|---------|----------|------------|--------|------------|------------|-------|----------|--------------|----------|-----

Or compute their horizontal positions:

In [21]:
hyades.where_in_sky(at=mtime,observer=tebas,inplace=True)
hyades

18 star(s):
|      |   MN |    HD |   HR |   HIP | Gl        | Name           | OtherDesignations                                                                     | ProperName     | Bayer   | Flamsteed   | Constellation   |   RAJ2000 |   DecJ2000 |   GalLonJ2000 |   GalLatJ2000 |   pmRA |   pmDec |   RadVel |   Distance |   Vmag |   Vmag_min |   Vmag_max |   B-V | SpType   |   Luminosity |   XJ2000 |   YJ2000 |   ZJ2000 |    VXJ2000 |    VYJ2000 |    VZJ2000 |   Primary | MultipleID   |   IsMultiple |   IsVariable |           tt |    jed |   RAJ2000t |   DecJ2000t |   RAEpoch |   DecEpoch |      HA |      az |      el |     zen |
|------|------|-------|------|-------|-----------|----------------|---------------------------------------------------------------------------------------|----------------|---------|-------------|-----------------|-----------|------------|---------------|---------------|--------|---------|----------|------------|--------|------------|------------|-------|--

And plot them again:

In [22]:
fig,axs = hyades.plot_stars(coords=['RAEpoch','DecEpoch'])
fig.savefig('gallery/hyades-precessed.png')
plt.close(fig) # Used only for README generation


<p align="center"><img src="https://github.com/seap-udea/MontuPython/blob/main/gallery/hyades-precessed.png?raw=true" alt="Logo""/></p>

### Working with planets and observing sites

`MontuPython` allows calculate the position of all planets in the solar system, including the moon:

In [23]:
mars = montu.Planet('Mars')
neptune = montu.Planet('neptune')
jupiter = montu.Planet('JUPITER')
moon = montu.Planet('Moon')
sun = montu.Planet('Sun')

You may create an observing site:

In [24]:
tebas = montu.Observer(lon=33,lat=24,height=0)

The straight routine method to calculate position of the planet in the sky at any date is:

In [25]:
mtime = montu.Time('-2500-01-01 12:00:00')
mars.where_in_sky(mtime,tebas)
mars.show_position()

Mars — sky position
  Epoch: -2500-01-22 12:00:00 / -2500-01-01 12:00:00.000000  (JED 807954.000000)
  Site: Custom site — lat 24.000000°, lon 33.000000°, 0 m  (P=1013.25 mbar, T=15.0 °C)
  Name: Mars
  RA (J2000): 12:31:51.677 h
  Dec (J2000): 1.620449°
  RA (epoch): 08:32:09.718 h
  Dec (epoch): 24.111367°
  RA (geocentric): 08:32:09.640 h
  Dec (geocentric): 24.114107°
  Azimuth: 6.189974°
  Elevation: -41.638524°


You may also compute other atrometric conditions (rise time, set time, elongation, etc.):

In [26]:
mars.conditions_in_sky(mtime,tebas)
mars.show_conditions()

Mars — sky conditions
  Epoch: -2500-01-22 12:00:00 / -2500-01-01 12:00:00.000000  (JED 807954.000000)
  Site: Custom site — lat 24.000000°, lon 33.000000°, 0 m  (P=1013.25 mbar, T=15.0 °C)
  Name: Mars
  Hour angle: 12:20:15.604 h
  Visual magnitude: -1.13 mag
  Rise time (UTC): -2500-01-22 16:49:49
  Rise azimuth: 63.133665°
  Set time (UTC): -2500-01-23 06:24:24
  Set azimuth: 296.924743°
  Transit time (UTC): -2500-01-22 23:37:37
  Transit elevation: 89.841412°
  Elongation from Sun: -157.818359°
  Distance from Earth: 0.660488 AU
  Distance from Sun: 1.626107 AU
  Angular diameter: 14.171"
  Illuminated fraction: 98.65 %
  Heliocentric latitude: 1.888422°
  Heliocentric longitude: 111.258645°
  Heliocentric longitude (alt.): 111.258645°
  Circumpolar: no
  Never rises: no


All times are given in Julian Days. If you want the exact date of set use:

In [27]:
montu.Time.get_date(mars.condition.rise_time)

(-2501, 1, 22, 16, 49, 57.292158)

You may get also the position at different times and store it into a data frame:

In [28]:
import numpy as np

# Reset the data stored in the planet
mars.reset_store()

# Loop on different times
for deltat in np.arange(0,1*montu.DAY,1*montu.HOUR):
    mars.where_in_sky(mtime + deltat,tebas,store=True)

# Show results
mars

Object Mars positions:
|            tt |    jed | Name   |   RAJ2000 |   DecJ2000 |   RAEpoch |   DecEpoch |   RAGeo |   DecGeo |        el |        az |
|---------------|--------|--------|-----------|------------|-----------|------------|---------|----------|-----------|-----------|
| -142006202700 | 807954 | Mars   |   12.531  |    1.62045 |   8.53603 |    24.1114 | 8.53601 |  24.1141 | -41.6385  |   6.18997 |
| -142006199100 | 807954 | Mars   |   12.5304 |    1.62593 |   8.53532 |    24.1153 | 8.53524 |  24.1179 | -38.0804  |  23.5034  |
| -142006195500 | 807954 | Mars   |   12.5297 |    1.63142 |   8.53461 |    24.1193 | 8.53447 |  24.1217 | -31.0231  |  37.8391  |
| -142006191900 | 807954 | Mars   |   12.5291 |    1.63692 |   8.53388 |    24.1233 | 8.53369 |  24.1256 | -21.5453  |  48.9456  |
| -142006188300 | 807954 | Mars   |   12.5284 |    1.64242 |   8.53314 |    24.1274 | 8.53292 |  24.1294 | -10.5149  |  57.4706  |
| -142006184700 | 807954 | Mars   |   12.5277 |    1.64793 |

### Working with the sun and the moon

There are two special planetary objects, the Sun and the Moon. Although you can define them as if they where planets:

In [29]:
sun = montu.Planet('Sun')
moon = montu.Planet('Moon')

and as we did with planets, you can calculate positions and conditions:

In [30]:
mtime = montu.Time('-2500-01-01 12:00:00')
Tebas = montu.Observer(lon=33,lat=24)
moon.conditions_in_sky(at=mtime,observer=Tebas)
moon.show_conditions()

Moon — sky conditions
  Epoch: -2500-01-22 12:00:00 / -2500-01-01 12:00:00.000000  (JED 807954.000000)
  Site: Custom site — lat 24.000000°, lon 33.000000°, 0 m  (P=1013.25 mbar, T=15.0 °C)
  Name: Moon
  Hour angle: 11:40:41.857 h
  Visual magnitude: -11.92 mag
  Rise time (UTC): -2500-01-22 17:53:53
  Rise azimuth: 67.113965°
  Set time (UTC): -2500-01-23 07:34:34
  Set azimuth: 289.880687°
  Transit time (UTC): -2500-01-23 00:47:47
  Transit elevation: 85.450065°
  Elongation from Sun: -148.449280°
  Distance from Earth: 0.002553 AU
  Distance from Sun: 0.997457 AU
  Angular diameter: 1879.512"
  Illuminated fraction: 92.65 %
  Heliocentric latitude: 5.076801°
  Heliocentric longitude: 133.820563°
  Heliocentric longitude (alt.): 133.820563°
  Circumpolar: no
  Never rises: no


There are additional methods, both proper and static, that can be used with these objects. For instance, you can:

- calculate the time of twilight:

In [31]:
dusk_time, dawn_time = montu.Sun.when_is_twilight(day=mtime,observer=Tebas,sunbelow=-18)
Tebas.get_local_time(dusk_time), Tebas.get_local_time(dawn_time) 

('05:34:16.262', '18:55:22.138')

- Get the date of solstices and equinoxes:

In [32]:
mtime = montu.Time('2023-01-01 12:00:00')
vernal, summer, autumn, winter = montu.Sun.next_seasons(at=mtime)

(
    montu.Time.get_date(vernal,format='mtime'),
    montu.Time.get_date(summer,format='mtime'),
    montu.Time.get_date(autumn,format='mtime'),
    montu.Time.get_date(winter,format='mtime'),
)

(Time('2023-03-20 21:25:28.5'/'2023-03-20 21:24:24'/'[hrw 4806] IV shemu 7'/JED 2460024.3918415/JTD 2460024.392691),
 Time('2023-06-21 14:59:09.1'/'2023-06-21 14:57:57'/'[hrw 4807] III akhet 5'/JED 2460117.123559/JTD 2460117.1244109),
 Time('2023-09-23 06:51:14.0'/'2023-09-23 06:50:50'/'[hrw 4807] II peret 9'/JED 2460210.7847257/JTD 2460210.7855787),
 Time('2023-12-22 03:28:23.1'/'2023-12-22 03:27:27'/'[hrw 4807] I shemu 9'/JED 2460300.6438576/JTD 2460300.6447118))

- Get the lunar phases:

In [33]:
montu.Moon.next_moon_quarters(since=mtime,output='datepro')

{'full': ['2023-01-06 23:09:10.797107', 7.907907465007156, 5.46470835711807],
 'last': ['2023-01-15 02:11:35.301130', 8.126672363374382, 13.591380720492452],
 'new': ['2023-01-21 20:54:23.100479', 6.7797199985943735, 20.371100719086826],
 'first': ['2023-01-28 15:20:01.694387',
  6.767808315809816,
  27.138909034896642]}

### An indepth examples

Short case studies: **Sirius heliacal rise** at the 139 CE Sothic apokatastasis, **Mars–Aldebarán** (September 2022), **Thales' eclipse** with Xavier Jubier cross-checks, and **polar stars** over 40 kyr.

### Heliacal rise of Sirius at the 139 CE apokatastasis


The Egyptian civil year begins on **I akhet 1**, when Sirius reappears in the dawn sky. The **139 CE apokatastasis** is one Sothic New Year in that cycle (`[hrw 2922] I akhet 1`). The `HeliacalRise` class (Schaefer 1987 model) predicts first-morning visibility of Sirius during the year anchored on that date:


In [34]:
tebas = montu.Observer(site="thebes")
sirius = montu.Stars(subset="bright", ProperName="Sirius")

calculator = montu.HeliacalRise(
    model="ptolemy",
    arcus_visionis_crit=14.0,  # assumed Arcus Visionis
)

t_start = montu.Time("139-06-01")
events = calculator.compute(
    sirius,
    tebas,
    t_start,
    t_start.add(years=1),
)

calculator.print_rises(
    events,
    title="Heliacal rises of Sirius (139 CE apokatastasis)",
    body_label="Sirius",
)


Heliacal rises of Sirius (139 CE apokatastasis) — 1 date(s)
  [1] 139-07-20 00:00:00  04:07:44.893  0139-07-19 00:00:00.000000  [hrw 2922] I akhet 1  Sirius -0.57°  Sun -14.18°
  source: Toomer, G. J. (1998). Ptolemy's Almagest. Princeton University Press. Book XIII, Chapter 7: "On the heliacal risings and settings of the planets".


The first heliacal rising falls on the same civil date as **I akhet 1** — the canonical anchor of the Egyptian year. See [`MontuPython-HeliacalRises.ipynb`](examples/MontuPython-HeliacalRises.ipynb) for a step-by-step treatment and alternative visibility models.


<p align="center"><img src="https://github.com/seap-udea/MontuPython/blob/main/montu_gui/assets/illustrations/solar-eclipses-illustration.webp?raw=true" alt="Illustration of a hypothetical total solar eclipse observed at Stonehenge" width="800"/></p>

<p align="center"><em>Illustration of a hypothetical total solar eclipse observed at Stonehenge</em>. Image generated with Nano Banana, edited in Preview.</p>

### Mars and Aldebarán Conjunction (September 2022)


The `Conjunction` and `ConjunctionExplorer` classes evaluate angular groupings of planets and stars. Here we recover the Mars–Aldebarán conjunction of 7 September 2022:


In [35]:
mars = montu.Planet("Mars")
aldebaran = montu.Stars(subset="bright", ProperName="Aldebaran", return_as="Star")
explorer = montu.ConjunctionExplorer(bodies=[mars, aldebaran], maxseparation=5)
conjs = explorer.search(
    start=montu.Time("2022-09-01"),
    end=montu.Time("2022-10-01"),
    observer="geocentric",
)
conjs[0].show_details()


Conjunction: Mars–Aldebaran
  Epoch (UTC)          : 2022-09-07 14:28:28
  Julian Day (UTC)     : 2459830.103152
  Observer             : geocentric
  Angular separation   : 4.2746° (max allowed 5.0°)
  In conjunction       : yes
  Is visible from site : n/a (geocentric)
  Pair Mars–Aldebaran
    Separation         : 4.2746°
    Position angle     : 170.19° (N→E)
  Mars
    Phase              : 85.42%
    Angular size       : 0.169 arcmin
    V magnitude        : -0.22
  Aldebaran
    V magnitude        : 0.87


See [`MontuPython-Conjunctions.ipynb`](examples/MontuPython-Conjunctions.ipynb) for lapse intervals, visibility checks, sky maps, and mixed planet–star groupings.


### Thales' eclipse (585 BCE)


The [NASA Five Millennium Canon of Solar Eclipses](https://eclipse.gsfc.nasa.gov/SEpubs/5MCSE.html) records the **total eclipse of 28 May 585 BCE** (catalogue year −584), associated with Thales of Miletus. MontuPython loads the catalogue row and computes local circumstances for any `Observer`.

We start by looking all eclipses in a given year interval:

In [36]:
catalogue = montu.SolarEclipses()
eclipses = catalogue.get_eclipses(year=[-590, -580], eclipse_type='T')
TABLEDF(eclipses.data)


|      |   year |   month |   day | td_ge    |      dt |   luna_num |   saros | eclipse_type   |    gamma |   magnitude | lat_ge   | lng_ge   |   lat_dd_ge |   lng_dd_ge |   sun_alt |   sun_azm |   path_width | central_duration   |   duration_secs |   cat_no |   canon_plate |   julian_date |   t0 |        x0 |       x1 |        x2 |        x3 |        y0 |         y1 |         y2 |        y3 |        d0 |        d1 |     d2 |      mu0 |     mu1 |   mu2 |      l10 |       l11 |       l12 |       l20 |       l21 |       l22 |    tan_f1 |    tan_f2 |   tmin |   tmax |   etype |   PNS |   UNS |   NCN |   nSer |   nSeq |   nJLE |
|------|--------|---------|-------|----------|---------|------------|---------|----------------|----------|-------------|----------|----------|-------------|-------------|-----------|-----------|--------------|--------------------|-----------------|----------|---------------|---------------|------|-----------|----------|-----------|-----------|-----------|---------

Let's load the particular eclipse we want to study:

In [37]:
eclipse = montu.SolarEclipse(eclipses.data.iloc[3])
print(eclipse)

SolarEclipse
Date (catalogue): -0584-05-28
Catalogue
  Eclipse type         : T (total)
  γ                    : 0.32013 R⊕
  magnitude            : 1.07977
  julian_date          : 1507900.31200 (JD TT)
  ΔT assumed           : 18383.9 s
  saros                : 57
  luna_num             : -31955
  cat_no               : 3379
Greatest eclipse
  td_ge (TT)           : 19:28:50
  lat_ge, lng_ge       : 38.2N, 45.0W
  lat_dd_ge            : 38.15594°
  lng_dd_ge            : -45.02063°
  sun_alt, sun_azm     : 71.1°, 158.5°
Central path
  path_width           : 271.5 km
  central_duration     : 06m04s
  duration_secs        : 364.2 s
  path_map             : http://xjubier.free.fr/en/site_pages/solar_eclipses/xSE_GoogleMap3.php?Ecl=-05840528&Acc=2&Umb=1&Lmt=1&Mag=0


In [38]:
troy = montu.Observer(site="troy")
conditions = eclipse.conditions_eclipse(troy)
conditions.show_details()

Eclipse local circumstances
  Catalogue date       : -0584-05-28 (T, total)
  Observer             : lat 39.957500°, lon 26.238900°, 30 m
  Kind                 : total
  Visible              : yes
  Magnitude            : 1.031
  Obscuration          : 1.000
  Moon/Sun radius ratio: 1.0667
  Sun altitude at max  : 14.77°
  Maximum (UTC)        : -584-05-28 15:56:56
  Maximum (JD UT)      : 1507900.164301
  Maximum (JD TT)      : 1507900.377078
  t_max                : 2.042425 h = 122.545470 min (from catalogue t0)
Contacts (UTC)
  C1 (first contact)   : -584-05-28 14:57:57 (alt 26.0°, az 275.5°)
  C2 (second contact)  : -584-05-28 15:54:54 (alt 15.1°, az 284.1°)
  C3 (third contact)   : -584-05-28 15:58:58 (alt 14.4°, az 284.7°)
  C4 (fourth contact)  : -584-05-28 16:51:51 (alt 4.8°, az 292.7°)
  Umbra duration       : 00:03:40
  cond_map             : http://xjubier.free.fr/en/site_pages/solar_eclipses/xSE_GoogleMap3.php?Ecl=-05840528&Acc=2&Umb=1&Lmt=1&Mag=0&Lat=39.9575&Lng=26.2389&

**Xavier Jubier cross-checks** — the `print` output above lists `path_map` and `cond_map` as plain text; use these links to open the interactive maps in a browser:

- [`path_map`](http://xjubier.free.fr/en/site_pages/solar_eclipses/xSE_GoogleMap3.php?Ecl=-05840528&Acc=2&Umb=1&Lmt=1&Mag=0) — central eclipse path for 28 May 585 BCE (catalogue year −584)
- [`cond_map`](http://xjubier.free.fr/en/site_pages/solar_eclipses/xSE_GoogleMap3.php?Ecl=-05840528&Acc=2&Umb=1&Lmt=1&Mag=0&Lat=39.9575&Lng=26.2389&Elv=30.0&Zoom=9&LC=1) — local circumstances at Troy (Ilion), the observer site in this example

Further case studies: [`MontuPython-SolarEclipses.ipynb`](examples/MontuPython-SolarEclipses.ipynb).


### Evolution of polar stars

Choose from database all bright stars that according to [wikipedia](https://en.wikipedia.org/wiki/Pole_star#Precession_of_the_equinoxes) were or will be close to the celestial North pole:

In [39]:
star_names = ('Polaris','Vega','Thuban','Deneb','Alderamin','Kochab')
stars = stars.get_stars(ProperName=star_names)
stars

6 star(s):
|     |   MN |     HD |   HR |    HIP | Gl     | Name      | OtherDesignations                                                                   | ProperName   | Bayer   | Flamsteed   | Constellation   |   RAJ2000 |   DecJ2000 |   GalLonJ2000 |   GalLatJ2000 |   pmRA |   pmDec |   RadVel |   Distance |   Vmag |   Vmag_min |   Vmag_max |    B-V | SpType       |   Luminosity |     XJ2000 |     YJ2000 |    ZJ2000 |    VXJ2000 |   VYJ2000 |    VZJ2000 |   Primary |   MultipleID |   IsMultiple |   IsVariable |
|-----|------|--------|------|--------|--------|-----------|-------------------------------------------------------------------------------------|--------------|---------|-------------|-----------------|-----------|------------|---------------|---------------|--------|---------|----------|------------|--------|------------|------------|--------|--------------|--------------|------------|------------|-----------|------------|-----------|------------|-----------|-------------

Now precess the position of all stars from -20 000 to 20 000 years from 2000:

In [40]:
import pandas as pd
import time

now = montu.Time()
epochs = np.linspace(-20000 * montu.YEAR, 20000 * montu.YEAR, 1000)
n = len(epochs)
print(f"Precessing {len(star_names)} stars over {n:,} epochs …")

t0 = time.perf_counter()
rows = []
for dt in epochs:
    past = now + dt
    pstars = stars.where_in_space(at=past)
    row = {"tt": past.tt}
    for star in star_names:
        row[star] = pstars.value_for(star, "DecEpoch")
    rows.append(row)

df = pd.DataFrame(rows)
print(f"Done in {time.perf_counter() - t0:.1f} s.")


Precessing 6 stars over 1,000 epochs …


Done in 1.4 s.


Now plot declinations as a function of time:

In [41]:
fig,ax = plt.subplots(figsize=(10,6))
for star in star_names:
    ax.plot(df['tt'],df[star],label=star)

ax.legend(loc='lower center',ncol=len(star_names))
ax.set_xlabel("Time [year]")
ax.set_ylabel("Declination [deg]")
ax.axvline(montu.Time().tt,lw=3)
ax.text(0.5,1.01,'Now',ha='center',transform=ax.transAxes)
ax.margins(0)
ax.set_xticks(np.linspace(df['tt'].min(),df['tt'].max(),14))
ax.grid()
montu.Time.set_time_ticks(ax)
montu.Util.montu_mark(ax)
fig.tight_layout()
fig.savefig('gallery/pole-stars.png')
plt.close(fig) # Used only for README generation


<p align="center"><img src="https://github.com/seap-udea/MontuPython/blob/main/gallery/pole-stars.png?raw=true" alt="Logo""/></p>

Check date when star is close to the pole:

In [42]:
for star in star_names:
    imax = df[star].argmax()
    mtime = montu.Time(df.iloc[imax].tt).get_readable()
    print(f"Star {star} will be the closest to the pole at {mtime.readable.datespice} (declination {montu.D2S(df.iloc[imax][star])})")

Star Polaris will be the closest to the pole at 2086-08-18 14:51:58.406394 (declination 89:31:55.872)
Star Vega will be the closest to the pole at 11609 B.C. 08-21 07:43:26.302144 (declination 86:22:03.668)
Star Thuban will be the closest to the pole at 2800 B.C. 08-23 15:03:14.495072 (declination 89:56:05.132)
Star Deneb will be the closest to the pole at 14732 B.C. 06-10 03:12:17.994176 (declination 86:57:15.599)
Star Alderamin will be the closest to the pole at 7532-03-09 08:41:33.198656 (declination 87:58:43.020)
Star Kochab will be the closest to the pole at 1078 B.C. 05-27 23:10:16.98240 (declination 83:29:32.452)


### Working with Observer Horizons

MontuPython can download real terrain data (Copernicus GLO-30 DEM, 30 m resolution) and compute the true visible horizon profile for any site on Earth. Here we show how the Sun rose through the **Royal Wadi** of ancient Akhetaten (modern Amarna) — the sacred notch that Akhenaten chose as the *akhet* ('horizon') symbol of his new city.

In [43]:
site = montu.Observer(site='amarna')
site.horizon_profile(
    max_dist=30,    # search radius in km
    az_step=0.5,    # azimuth resolution in degrees
    coarse_step=0.1 # radial scan step in km
)
print(site.horizon)

Obtaining horizon profile...


Horizon for 'Amarna (Akhetaten)'
  Coordinates: lat=27.6444, lon=30.9014, alt=90 m
  Status: computed (720 pts)
  Elevation range: [-0.22°, 1.22°]
  Parameters: max_dist=30 km, az_step=0.5°, coarse_step=0.1 km


The `horizon_profile()` call downloads and caches the DEM tiles automatically. Once computed, `site.horizon.get_elevation(az)` returns the terrain elevation angle at any azimuth.

We can now refine the flat-horizon sunrise time to account for the real terrain:

In [44]:
sun = montu.Sun()
mtime = montu.Time('bce1341-10-21')
sun.conditions_in_sky(at=mtime, observer=site, horizon=True)

mtime_rise_hor = montu.Time(sun.condition.rise_time_hor, format='jd')
print(f"Flat-horizon rise: {site.get_local_time(sun.condition.rise_time)}")
print(f"Terrain-corrected: {site.get_local_time(sun.condition.rise_time_hor)}")
print(f"Rise azimuth (terrain): {sun.condition.rise_az_hor:.2f}°")

Flat-horizon rise: 06:10:12.448
Terrain-corrected: 06:15:42.721
Rise azimuth (terrain): 102.94°


The terrain correction shifts the sunrise by a few minutes because the Sun must climb above the wadi walls before it becomes visible.

Finally, we can plot the sky at the horizon-corrected sunrise moment, showing the Sun emerging from the Royal Wadi:

In [45]:
fig = site.horizon.plot_horizon(
    at=mtime_rise_hor,
    az_center=90,
    az_delta=40,
    elev_view=4,
    mag_limit=-1,
    show_constellation_lines=False,
    show_star_names=False,
    show_planets=['Sun'],
)

See [`MontuPython-ObserverHorizon.ipynb`](examples/MontuPython-ObserverHorizon.ipynb) for a full walkthrough including multi-site comparisons, Valley of the Kings, the Senenmut temple, and the north shaft of the Khufu pyramid.



------------

This package has been designed and written by [Jorge I. Zuluaga](https://jorgezuluaga.github.io) (C) 2023-present